Generate stub/stripped Any*Event enums in event_enum! macro

DevinR528 opened the pull request:
I didn't want to push this to the other PR incase it wasn't the direction you were thinking, if it is I can either merge it into the first PR or close that one. I also took a shot at handling the State, Message, and Basic event deserialization when a custom event is encountered.

jimmycuadra commented:
Thanks for the PR! The spec doesn't say anything about what deactivating means other than "removing all ability for the user to login again." If the user record were to be deleted from the database, that might be a problem since there will still be events in the graph for any rooms they joined that reference them. Given the limited scope of being able to log in that the spec describes, I think you're right that we should just have an "active" field for users that is true initially and stays true until the deactivate. Then we'd need to check the value of this field before issuing any new access tokens for the user.

erikjohnston commented:
(The spec may also want to mention that the API should probably also delete
any personal information associated with the account, e.g. email addresses
etc, fwiw)

On Sat, 17 Sep 2016 19:43 Jimmy Cuadra, <notifications@github.com> wrote:

> Thanks for the PR! The spec doesn't say anything about what deactivating
> means other than "removing all ability for the user to login again." If the
> user record were to be deleted from the database, that might be a problem
> since there will still be events in the graph for any rooms they joined
> that reference them. Given the limited scope of being able to log in that
> the spec describes, I think you're right that we should just have an
> "active" field for users that is true initially and stays true until the
> deactivate. Then we'd need to check the value of this field before issuing
> any new access tokens for the user.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/ruma/ruma/pull/89#issuecomment-247796759>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AICaWMbBznQFL8quqffDjOWV1EbCg_RKks5qrDTJgaJpZM4J_eL0>
> .
>

jplatte commented:
Shouldn't the endpoint be called `AccountDeactivate` though, to stay consistent with the existing naming scheme?

mujx commented:
I updated the PR using the `active` flag. What would be the process of reactivating the account?
I couldn't find anything relevant in the spec. 

Maybe the `POST /_matrix/client/r0/account/password/email/requestToken` endpoint?

@jplatte I think it would be a little awkward to read. Also the other endpoints use the verb first,
which is more natural. The `AccountPassword` could be renamed to `ChangeAccountPassword`.

jplatte commented:
The other endpoints don't have a descriptive verb first, they have the HTTP method first (in case the URL supports multiple HTTP methods). I agree that if the endpoints should be reflective of what they do instead of their http method / url, changing `AccountPassword` to `ChangeAccountPassword` or similar would make sense. But I guess the naming scheme is up to @jimmycuadra to decide.

jimmycuadra commented:
My guess is that once the account is deactivated it is gone for good. This seems supported by Erik's comment that personal information should be removed from the user when deactivated.

Everyone is right that the struct names for the API endpoints are a bit awkward right now. You can leave them as is here and I'll change them all at once afterwards.

mujx commented:
- The `user/{userId}/account_data/{type}` endpoint was added. #63 
- The `deactivate` endpoint was updated to delete the account data.

I think that was the info @erikjohnston was referring to.
The account data per room has similar implementation, and it will be added soon. #62 

Let me know if you have any improvements/changes in mind.

jimmycuadra commented:
Looks great! Thanks so much for your work on this.

mujx commented:
Thanks for merging!

DevinR528 committed:
Add Any*EventStub and AnyStrippedStateEventStub to enums.rs

Conditionally emit tokens for content enum in event_enum! and the path
for each Any*Event variant contents. Add tests for redaction events now
that they are part of Any*Event enums. Fix any tests that used
Any*EventContent.

DevinR528 committed:
Any*Event enums handle custom events instead of returning an error

DevinR528 committed:
Use existing event_enum! macro invocations to generate Stub/Stripped

jplatte commented on ruma-events/src/enums.rs:
```diff
@@ -66,7 +60,7 @@ event_enum! {
         "m.room.pinned_events",
         "m.room.power_levels",
-        "m.room.redaction",
+        // "m.room.redaction",
```
?

DevinR528 replied:
This needs to be removed right?

jplatte replied:
Oh, I didn't notice that redactions were part of `AnyStateEvent`. Yes, remove this line completely.

jplatte replied:
Maybe also remove the empty line at the end of `AnyMessageEvent`s event list ^^

But I'm glad it's not missing there.

jplatte commented on ruma-events/src/enums.rs:
```diff
@@ -144,7 +128,7 @@ impl<'de> de::Deserialize<'de> for AnyEvent {
     {
         let json = Box::<RawJsonValue>::deserialize(deserializer)?;
-        let EventDeHelper { ev_type } = from_raw_json_value(&json)?;
+        let EventDeHelper { ev_type, state_key, event_id } = from_raw_json_value(&json)?;
```
Nice! Though if you're doing this, you should probably get rid of `is_compatible` at the same time (known event_type strings don't take precedence over `event_id` / `state_key` being present in the event fields).

DevinR528 replied:
Would adding a room_id to the `EventDeHelper` then solve differentiating `Ephemeral` and `Baisc` events?

jplatte replied:
Yes! :)

jplatte replied:
Now that I'm thinking about it: `EventDeHelper` should use `Option<IgnoredAny>` instead of the actual field types that will later be used. We only care whether the fields are present, not what their values are.

DevinR528 replied:
Good thinking!

jplatte commented on ruma-events/src/lib.rs:
```diff
@@ -240,8 +240,11 @@ pub trait StateEventContent: RoomEventContent {}
+    #[serde(default)]
+    pub event_id: Option<String>,
 }
-
```
?

jplatte commented on ruma-events/src/lib.rs:
```diff
@@ -240,8 +240,11 @@ pub trait StateEventContent: RoomEventContent {}
 pub struct EventDeHelper {
     #[serde(rename = "type")]
     pub ev_type: String,
+    #[serde(default)]
```
`default` is not required for `Option`.

DevinR528 committed:
Remove redaction event from state events, handle ephemeral events in Any*Event enums

jplatte commented on ruma-events/src/enums.rs:
```diff
@@ -128,33 +126,18 @@ impl<'de> de::Deserialize<'de> for AnyEvent {
-            }
+        let EventDeHelper { state_key, event_id, room_id, .. } = from_raw_json_value(&json)?;
+
+        // Determine wether the event is a state, message, ephemeral, or basic event
```
```suggestion
        // Determine whether the event is a state, message, ephemeral, or basic event
```

jplatte commented on ruma-events/src/lib.rs:
```diff
@@ -234,17 +237,24 @@ pub trait MessageEventContent: RoomEventContent {}
+    /// If present the event is a state event.
+    pub state_key: Option<IgnoredAny>,
+
+    /// If no `state_key` is found but an `event_id` is present the a message event is deserialized.
```
Please re-read and make the comments of `state_key`, `event_id` and `room_id` consistent which each other.

DevinR528 committed:
Clarify/fix EventDeHelper docs
